File: java-version/src/main/java/com/example/java_version/service/CustomUserDetailsService.java
package com.example.java_version.service;

import com.example.java_version.repository.SampleUserRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.core.userdetails.UserDetailsService;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.stereotype.Service;

@Service
public class CustomUserDetailsService implements UserDetailsService {

    @Autowired
    private SampleUserRepository userRepository;

    @Override
    public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException {
        return userRepository.findUserByEmail(username).orElseThrow(
                () -> new UsernameNotFoundException("Usuário não encontrado!"));
    }
}

File: kotlin-version/src/main/kotlin/com/example/kotlin_version/service/CustomUserDetailsService.kt
package com.example.kotlin_version.service

import com.example.kotlin_version.repository.SampleUserRepository
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.security.core.userdetails.UserDetailsService
import org.springframework.security.core.userdetails.UsernameNotFoundException
import org.springframework.stereotype.Service

@Service
class CustomUserDetailsService(
    private val userRepository: SampleUserRepository
): UserDetailsService {

    override fun loadUserByUsername(username: String): UserDetails {
        return userRepository.findUserByEmail(username) ?: throw UsernameNotFoundException("Usuário não encontrado!")
    }
}